flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -312,20 +347,104 @@ public void initializeState(FunctionInitializationContext context) throws Except
 		}
 		this.pendingCommitTransactions.clear();
 
-		currentTransaction = beginTransaction();
-		LOG.debug("{} - started new transaction '{}'", name(), currentTransaction);
+		currentTransactionHolder = beginTransactionInternal();
+		LOG.debug("{} - started new transaction '{}'", name(), currentTransactionHolder);
+	}
+
+	/**
+	 * This method must be the only place to call {@link #beginTransaction()} to ensure that the
+	 * {@link TransactionHolder} is created at the same time.
+	 */
+	private TransactionHolder<TXN> beginTransactionInternal() throws Exception {
+		return new TransactionHolder<>(beginTransaction(), clock.millis());
+	}
+
+	/**
+	 * This method must be the only place to call {@link #recoverAndCommit(Object)} to ensure that
+	 * the configuration parameters {@link #transactionTimeout} and
+	 * {@link #ignoreFailuresAfterTransactionTimeout} are respected.
+	 */
+	private void recoverAndCommitInternal(TransactionHolder<TXN> transactionHolder) {
+		try {
+			logWarningIfTimeoutAlmostReached(transactionHolder);
+			recoverAndCommit(transactionHolder.handle);
+		} catch (final Exception e) {
```

Comment:
@GJL @pnowojski I came across this method today and was initially confused that this is actually possible, i.e. catching and then possibly re-throwing a checked `Exception` without mentioning it in the method signature. But apparently, this is valid Java as a special case of `Exception` handling where the checked `Exception` is actually a `RuntimeException`.

That said, this `catch` block should actually focus on `RuntimeExceptions` because `logWarningIfTimeoutAlmostReached` and `recoverAndCommit` both are not considering checked `Exceptions`. I'm just asking because I got curious whether I miss something here and this `catch` was done intentionally or whether this minor thing just slipped through...

Reply:
I think changing `Exception` to `RuntimeException` won't be semantically equivalent due to the possibility of [sneaky throws](https://www.baeldung.com/java-sneaky-throws#sneaky). My memory about this code is a bit fuzzy but I assume there can be sneaky throws from within `recoverAndCommit` because the `commit` callback is public API.

Reply:
Cool thanks for clarification. That makes sense. And thanks for the pointer to the sneaky throw example. 👍